From 83711fb355b21c5e8f8792bcf591cf677dae81b8 Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Tue, 18 Feb 2014 23:39:34 +0000
Subject: [PATCH 30/31] Add {e}glibc 2.15 patch for confstr.c

The original commit for this was done so badly that I will not
provide direct links to the commits; you can trace them from
https://dmz-portal.mips.com/bugz/show_bug.cgi?id=426 though.
---
 ...tr-use-of-local-buffer-outside-its-extent.patch | 55 ++++++++++++++++++++++
 ...tr-use-of-local-buffer-outside-its-extent.patch | 35 ++++++++++++++
 2 files changed, 90 insertions(+)
 create mode 100644 patches/eglibc/2_15/180-Fix-confstr-use-of-local-buffer-outside-its-extent.patch
 create mode 100644 patches/glibc/2.15/180-Fix-confstr-use-of-local-buffer-outside-its-extent.patch

diff --git a/patches/eglibc/2_15/180-Fix-confstr-use-of-local-buffer-outside-its-extent.patch b/patches/eglibc/2_15/180-Fix-confstr-use-of-local-buffer-outside-its-extent.patch
new file mode 100644
index 0000000..25b7f93
--- /dev/null
+++ b/patches/eglibc/2_15/180-Fix-confstr-use-of-local-buffer-outside-its-extent.patch
@@ -0,0 +1,55 @@
+diff -urN a/posix/confstr.c b/posix/confstr.c
+--- a/posix/confstr.c	2014-02-19 00:13:49.954666882 +0000
++++ b/posix/confstr.c	2014-02-19 00:19:02.019492923 +0000
+@@ -36,6 +36,10 @@
+   const char *string = "";
+   size_t string_len = 1;
+ 
++  /* Note that this buffer must be large enough for the longest strings
++     used below.  */
++  char restenvs[4 * sizeof "POSIX_V7_LPBIG_OFFBIG"];
++
+   switch (name)
+     {
+     case _CS_PATH:
+diff -urN a/posix/confstr.inc b/posix/confstr.inc
+--- a/posix/confstr.inc	2014-02-19 00:13:49.958000267 +0000
++++ b/posix/confstr.inc	2014-02-19 00:17:22.854625641 +0000
+@@ -29,7 +29,6 @@
+ 
+ 	 Currently this means all environment which the system allows.  */
+       {
+-	char restenvs[4 * sizeof "POSIX_V7_LPBIG_OFFBIG"];
+ 
+ 	string_len = 0;
+ #ifndef _POSIX_V7_ILP32_OFF32
+@@ -92,7 +91,6 @@
+ 
+ 	 Currently this means all environment which the system allows.  */
+       {
+-	char restenvs[4 * sizeof "POSIX_V6_LPBIG_OFFBIG"];
+ 
+ 	string_len = 0;
+ #ifndef _POSIX_V6_ILP32_OFF32
+@@ -155,7 +153,6 @@
+ 
+ 	 Currently this means all environment which the system allows.  */
+       {
+-	char restenvs[4 * sizeof "XBS5_LPBIG_OFFBIG"];
+ 
+ 	string_len = 0;
+ #ifndef _XBS5_ILP32_OFF32
+diff -urN a/posix/cross-getconf.c b/posix/cross-getconf.c
+--- a/posix/cross-getconf.c	2014-02-19 00:13:49.954666882 +0000
++++ b/posix/cross-getconf.c	2014-02-19 00:17:22.854625641 +0000
+@@ -93,6 +93,10 @@
+   const char *string = "";
+   size_t string_len = 1;
+ 
++  /* Note that this buffer must be large enough for the longest strings
++     used below.  */
++  char restenvs[4 * sizeof "POSIX_V7_LPBIG_OFFBIG"];
++
+   switch (name)
+     {
+ #include "confstr.inc"
diff --git a/patches/glibc/2.15/180-Fix-confstr-use-of-local-buffer-outside-its-extent.patch b/patches/glibc/2.15/180-Fix-confstr-use-of-local-buffer-outside-its-extent.patch
new file mode 100644
index 0000000..04e2a0b
--- /dev/null
+++ b/patches/glibc/2.15/180-Fix-confstr-use-of-local-buffer-outside-its-extent.patch
@@ -0,0 +1,35 @@
+diff -urN a/posix/confstr.c b/posix/confstr.c
+--- a/posix/confstr.c	2012-01-01 12:16:32.000000000 +0000
++++ b/posix/confstr.c	2014-02-18 23:35:58.722876213 +0000
+@@ -35,6 +35,7 @@
+ {
+   const char *string = "";
+   size_t string_len = 1;
++  char restenvs[4 * sizeof "POSIX_V7_LPBIG_OFFBIG"];
+ 
+   switch (name)
+     {
+@@ -55,7 +56,6 @@
+ 
+ 	 Currently this means all environment which the system allows.  */
+       {
+-	char restenvs[4 * sizeof "POSIX_V7_LPBIG_OFFBIG"];
+ 
+ 	string_len = 0;
+ #ifndef _POSIX_V7_ILP32_OFF32
+@@ -118,7 +118,6 @@
+ 
+ 	 Currently this means all environment which the system allows.  */
+       {
+-	char restenvs[4 * sizeof "POSIX_V6_LPBIG_OFFBIG"];
+ 
+ 	string_len = 0;
+ #ifndef _POSIX_V6_ILP32_OFF32
+@@ -181,7 +180,6 @@
+ 
+ 	 Currently this means all environment which the system allows.  */
+       {
+-	char restenvs[4 * sizeof "XBS5_LPBIG_OFFBIG"];
+ 
+ 	string_len = 0;
+ #ifndef _XBS5_ILP32_OFF32
-- 
1.9.0

